### PR TITLE
Use select when writing or reading from a channel

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

This PR is a follow up of our [Discuss](https://discuss.thethingsindustries.com/t/eu-pcn-optimizations-and-changes/255/3?u=adriansmares) topic on concurrency optimizations and correctness. After the PB [router](https://github.com/packetbroker/router/commit/39366f3b2ca7aa0ebfd110416a0bc038388184d4) bug and the leak I found this morning in the GS bridges I've searched through all of usages of `<-` and tried to make them context aware.

#### Changes

- Read or write from channels with `select`, binding to the context.

#### Testing

Unit testing.

##### Regressions

Shouldn't be the case.

#### Notes for Reviewers

`pkg/util/test` didn't seem to require this type of behavior, but we can update that one as well.

GS leak source:
image

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/applicationserver/io/io.go [hunk 1]
```diff
@@ -93,13 +93,14 @@ var errBufferFull = errors.DefineResourceExhausted("buffer_full", "buffer is ful
 // SendUp sends an upstream message.
 // This method returns immediately, returning nil if the message is buffered, or with an error when the buffer is full.
 func (s *Subscription) SendUp(ctx context.Context, up *ttnpb.ApplicationUp) error {
+	ctxUp := &ContextualApplicationUp{
```
Comment (adriansmares) [inline pkg/applicationserver/io/io.go:96]: Changed for consistency.

File: pkg/applicationserver/io/packages/packages.go [hunk 2]
```diff
@@ -142,7 +142,7 @@ func (s *server) NewSubscription() *io.Subscription {
 	go func() {
 		for {
 			select {
-			case <-s.ctx.Done():
+			case <-sub.Context().Done():
```
Comment (adriansmares) [inline pkg/applicationserver/io/packages/packages.go:145]: The `Up()` channel used before is available as long as the `sub` is alive (`sub.Context().Done()` is not closed)

File: pkg/gatewayserver/gatewayserver.go [hunk 3]
```diff
@@ -665,7 +667,11 @@ func (gs *GatewayServer) handleUpstream(conn connectionEntry) {
 						handleFn(ctx, host)
 					}()
 					go func() {
-						host.handleCh <- item
+						select {
+						case <-ctx.Done():
+							return ctx.Err()
+						case host.handleCh <- item:
```
Comment (adriansmares) [inline pkg/gatewayserver/gatewayserver.go:673]: This is the small part of the `bridge-4` leak.

File: pkg/gatewayserver/io/mqtt/mqtt.go [hunk 4]
```diff
@@ -124,12 +124,20 @@ func (c *connection) setup(ctx context.Context) (err error) {
...
+				}
 				return
 			}
 			if pkt != nil {
 				logger.Debugf("Schedule %s packet", packet.Name[pkt.PacketType()])
-				controlCh <- pkt
+				select {
+				case <-ctx.Done():
+					return
+				case controlCh <- pkt:
```
Comment (adriansmares) [inline pkg/gatewayserver/io/mqtt/mqtt.go:137]: This is the big part of the `bridge-4` leak.

**Review by htdvisser — APPROVED** [review]
Looks good to me 👍

**Review by KrishnaIyer — APPROVED** [review]
Code owned files (GS) LGTM.

**Review by KrishnaIyer — APPROVED** [review]
AS LGTM

**Review by johanstokking — COMMENTED** [review]
Nice!

**Pull request merged**